x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/async/AsyncTaskMaintenanceService.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.core.async;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.ClusterStateListener;
+import org.elasticsearch.cluster.routing.IndexRoutingTable;
+import org.elasticsearch.common.lease.Releasable;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.common.util.concurrent.EsRejectedExecutionException;
+import org.elasticsearch.gateway.GatewayService;
+import org.elasticsearch.index.query.QueryBuilders;
+import org.elasticsearch.index.reindex.DeleteByQueryAction;
+import org.elasticsearch.index.reindex.DeleteByQueryRequest;
+import org.elasticsearch.threadpool.Scheduler;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static org.elasticsearch.xpack.core.async.AsyncTaskIndexService.EXPIRATION_TIME_FIELD;
+
+/**
+ * A service that runs a periodic cleanup over the async execution index.
+ */
+public abstract class AsyncTaskMaintenanceService implements Releasable, ClusterStateListener {
+    private static final Logger logger = LogManager.getLogger(AsyncTaskMaintenanceService.class);
+
+    private final String index;
+    private final String localNodeId;
+    private final ThreadPool threadPool;
+    private final AsyncTaskIndexService<?> indexService;
+    private final TimeValue delay;
+
+    private boolean isCleanupRunning;
+    private final AtomicBoolean isClosed = new AtomicBoolean(false);
+    private volatile Scheduler.Cancellable cancellable;
+
+    public AsyncTaskMaintenanceService(String index,
```

Comment:
should it be the index service that provides that index name rather than having to provide it to both?